neo/Ledger/MemoryPool.cs

```diff
@@ -4,161 +4,357 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
-using System.Numerics;
+using System.Runtime.CompilerServices;
+using Neo.Persistence;
 
 namespace Neo.Ledger
 {
     internal class MemoryPool : IReadOnlyCollection<Transaction>
     {
-        private class PoolItem
+        private class PoolItem : IComparable
         {
             public readonly Transaction Transaction;
             public readonly DateTime Timestamp;
+            public readonly Fixed8 FeePerByte;
 
             public PoolItem(Transaction tx)
             {
                 Transaction = tx;
                 Timestamp = DateTime.UtcNow;
+                FeePerByte = Transaction.NetworkFee / Transaction.Size;
+            }
+
+            public int CompareTo(Transaction tx, Fixed8 feePerByte)
+            {
+                if (tx == null) return 1;
+                int ret = FeePerByte.CompareTo(feePerByte);
+                if (ret != 0) return ret;
+                ret = Transaction.NetworkFee.CompareTo(tx.NetworkFee);
+                if (ret != 0) return ret;
+
+                return Transaction.Hash.CompareTo(tx.Hash);
+            }
+            
+            public int CompareTo(object obj)
+            {
+                if (!(obj is PoolItem otherItem)) return 1;
+                return CompareTo(otherItem.Transaction, otherItem.FeePerByte);
             }
         }
 
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_fee = new ConcurrentDictionary<UInt256, PoolItem>();
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_free = new ConcurrentDictionary<UInt256, PoolItem>();
+        private static readonly double MaxSecondsToReverifyHighPrioTx = (double) Blockchain.SecondsPerBlock / 3;
+        private static readonly double MaxSecondsToReverifyLowPrioTx = (double) Blockchain.SecondsPerBlock / 5;
+        
+        // These two are not expected to be hit, they are just safegaurds. 
+        private static readonly double MaxSecondsToReverifyHighPrioTxPerIdle = (double) Blockchain.SecondsPerBlock / 15;
+        private static readonly double MaxSecondsToReverifyLowPrioTxPerIdle = (double) Blockchain.SecondsPerBlock / 30;
```

Comment:
15/3 = 5  15/5=3
15/15=1  15/30=0.5
safeguards are intended to be smaller?

Reply:
Yes the safeguards only validate like 10 messages at a time; should never take even close to a second or half a second; they are only used when validating messages while Blockchain is idle. They just ensure the blockchain actor is never stuck busy for a long time.

Note on this PR: I still have an additional change coming shortly to make the Idle message re-queue itself while there are still more unconfirmed transactions.